### DISCUSSION TITLE
Extensibility of Default Linker

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1770] [Q&A]
Hi, is there a reasoning why 

https://github.com/eclipse-langium/langium/blob/4e9681450f2dbc819b06a70b585d3d510066451d/packages/langium/src/references/linker.ts#L70
and
https://github.com/eclipse-langium/langium/blob/4e9681450f2dbc819b06a70b585d3d510066451d/packages/langium/src/references/linker.ts#L72

we would like to override and trap into

https://github.com/eclipse-langium/langium/blob/4e9681450f2dbc819b06a70b585d3d510066451d/packages/langium/src/references/linker.ts#L158

like

```
 get ref() {
        if (isAstNode(this._ref)) {
          // Most frequent case: the target is already resolved.
          const document = AstUtils.findRootNode(node).$document
          if (
            document !== undefined &&
            document.state >= DocumentState.Linked &&
            document.state < DocumentState.Validated
          ) {
            const sourceUri = document.uri.toString()
            const targetUri = AstUtils.getDocument(this._ref).uri.toString()
            if (sourceUri.endsWith(".someextension") && sourceUri !== targetUri) {
              mimimi()
            }
          }

          return this._ref
        } 
```

but currently this would require to c&p the complete class

**Comment by cdietrich** [comment]
@msujew any idea?

**Comment by msujew** [comment]
Hey @cdietrich,

Ah, right. I didn't really think about that. I'm wondering what the best solution to this is. @spoenemann any recommendations?

**Reply by spoenemann** [reply]
Hmm maybe something like this?

```typescript
export class DefaultLinker implements Linker {

    protected refAccessCallback?: (target: AstNode, ref: Reference) => void;

    ...

                    if (isAstNode(this._ref)) {
                        // Most frequent case: the target is already resolved.
                        if (linker.refAccessCallback) {
                            linker.refAccessCallback(this._ref, this);
                        }
                        return this._ref;
                    } ...

}
```

I'd like to keep the overhead for the default case (no callback configured) as low as possible.
